inspect-js-test.js

```diff
@@ -0,0 +1,40 @@
+
+const fs = require("fs")
+const glob = require("glob")
+const opaInspect = require("@zregvart/opa-inspect")
+
+function addCommas(jsonLines) {
+  return jsonLines.trim().split("\n").join(",")
+}
+
+function readFile(fileName) {
+  return fs.readFileSync(fileName).toString()
+}
+
+function readRegoAnnotations() {
+  var result = null
+
+  opaInspect(inspect => {
+    console.log("Inspecting rego...")
+
+    // We get JSONL style lines so that's why the addCommas is needed here
+    allJson = glob.sync("policy/**/*.rego").map(f => addCommas(inspect(f, readFile(f))))
+
+    result = JSON.parse(`{"annotations":[ ${allJson.join(",")} ]}`)
+
+    console.log("Real output is now ready...")
+
+    // This eventually shows the data we really want...
+    console.log(JSON.stringify(result.annotations.slice(0,2), null, 2))
+  })
+
+  // ..but this returns immediately so we always return null
+  // Need to await somehow but I've made so many different attempts
```

Comment:
With the existing state of `opaInspect`, if you want to create a wrapper that does something with the results, I believe you'll have to use some sort of callback:

```diff
diff --git a/inspect-js-test.js b/inspect-js-test.js
index 9fdc2ae..19977ec 100644
--- a/inspect-js-test.js
+++ b/inspect-js-test.js
@@ -11,7 +11,7 @@ function readFile(fileName) {
   return fs.readFileSync(fileName).toString()
 }
 
-function readRegoAnnotations() {
+function readRegoAnnotations(cb) {
   var result = null
 
   opaInspect(inspect => {
@@ -26,6 +26,8 @@ function readRegoAnnotations() {
 
     // This eventually shows the data we really want...
     console.log(JSON.stringify(result.annotations.slice(0,2), null, 2))
+
+    cb(result);
   })
 
   // ..but this returns immediately so we always return null
@@ -37,4 +39,6 @@ function readRegoAnnotations() {
   return result
 }
 
-console.log(`Result: ${readRegoAnnotations()}`)
+readRegoAnnotations(result => {
+    console.log(`Result: ${result}`)
+})
```

Or, maybe you can just call `opaInspect` directly with the function that you want. Any reason not to do that?

I think a better solution is probably to make `opaInspect` return a promise that you could chain on, e.g.:
```
opaInspect().then(...).then(...);
```

I don't think we should care too much that the line after `opaInspect` is executed before the data is actually read. If there's anything that relies on the result from opaInspect, use one of the approaches above.

Reply:
I've reworked the [opa-inspect-js](https://github.com/zregvart/opa-inspect-js), it now returns a Promise, so chaining should be possible.